src/transforms/pipelines/mod.rs

```diff
@@ -0,0 +1,200 @@
+mod expander;
+mod router;
+
+use crate::config::{
+    DataType, ExpandType, GenerateConfig, TransformConfig, TransformContext, TransformDescription,
+};
+use crate::transforms::Transform;
+use indexmap::IndexMap;
+use serde::{Deserialize, Serialize};
+
+const SAMPLE_CONFIG: &str = indoc::indoc! { r#"
+[logs]
+order = ["foo", "bar"]
+
+[logs.pipelines.foo]
+name = "foo pipeline"
+
+[[logs.pipelines.foo.transforms]]
+type = "filter"
+condition = ""
+
+[[logs.pipelines.foo.transforms]]
+type = "filter"
+condition = ""
+
+[logs.pipelines.bar]
+name = "bar pipeline"
+
+[[logs.pipelines.bar.transforms]]
+type = "filter"
+condition = ""
+"# };
+
+inventory::submit! {
+    TransformDescription::new::<PipelinesConfig>("pipelines")
+}
+
+#[derive(Debug, Default, Deserialize, Serialize)]
+// #[serde(deny_unknown_fields)]
+pub struct PipelineConfig {
+    name: String,
+    transforms: Vec<Box<dyn TransformConfig>>,
+}
+
+impl Clone for PipelineConfig {
+    fn clone(&self) -> Self {
+        // This is a hack around the issue of cloning
+        // trait objects. So instead to clone the config
+        // we first serialize it into JSON, then back from
+        // JSON. Originally we used TOML here but TOML does not
+        // support serializing `None`.
```

Comment:
This should already be possible, since we require that `TransformConfig` is a supertrait of `dyn_clone::DynClone`, and we use the `dyn_clone::clone_trait_object!` macro to define a `Clone` impl for `Box<dyn TransformConfig>`: https://github.com/vectordotdev/vector/blob/d849e22f8154442f276757f597e374673b67fb00/lib/vector-core/src/transform/config.rs#L45-L69

Should be as simple as deriving `Clone` on `PipelineConfig`.  Does doing that return a specific error?

Reply:
It's how it is currently done in the `ConfigBuilder` struct. I didn't want to reinvent the wheel on this.
Maybe we should create an issue related to this and handle it in a separate PR.

Reply:
I created this issue https://github.com/vectordotdev/vector/issues/9898